File: (行为)08.访问着模式.v2/golang/gpt/index.访问者.go
package main

import "fmt"

// Visitor 接口
type Visitor interface {
	VisitConcreteElementA(*ConcreteElementA)
	VisitConcreteElementB(*ConcreteElementB)
}

// Element 接口
type Element interface {
	Accept(Visitor)
}

// ConcreteElementA 结构体
type ConcreteElementA struct{}

func (e *ConcreteElementA) Accept(v Visitor) {
	v.VisitConcreteElementA(e)
}

// ConcreteElementB 结构体
type ConcreteElementB struct{}

func (e *ConcreteElementB) Accept(v Visitor) {
	v.VisitConcreteElementB(e)
}

// ConcreteVisitor 结构体
type ConcreteVisitor struct{}

func (v *ConcreteVisitor) VisitConcreteElementA(e *ConcreteElementA) {
	fmt.Println("访问者访问元素A")
}

func (v *ConcreteVisitor) VisitConcreteElementB(e *ConcreteElementB) {
	fmt.Println("访问者访问元素B")
}

// ObjectStructure 结构体，可以遍历 Element
type ObjectStructure struct {
	elements []Element
}

func (o *ObjectStructure) Add(element Element) {
	o.elements = append(o.elements, element)
}

func (o *ObjectStructure) Accept(v Visitor) {
	for _, e := range o.elements {
		e.Accept(v)
	}
}

/**
Visitor 接口：定义一个访问方法，该方法为每一种类型的被访问元素提供了一个访问的方法签名。
Element 接口：定义一个 Accept 方法，该方法应接收一个访问者对象。
ConcreteVisitor 类：实现 Visitor 接口的类，里面包含了对每一个元素类的访问逻辑。
ConcreteElement 类：实现 Element 接口的类，代表被访问的对象。
*/
func main() {
	// 创建ObjectStructure
	objects := &ObjectStructure{}

	// 向ObjectStructure添加元素
	objects.Add(&ConcreteElementA{})
	objects.Add(&ConcreteElementB{})

	// 创建访问者
	visitor := &ConcreteVisitor{}

	// 使用访问者访问对象结构中的元素
	objects.Accept(visitor)
	/**
	访问者访问元素A
	访问者访问元素B
	*/
}

File: (行为)08.访问着模式.v2/php/gpt/index.访问者.php
<?php
namespace php\gpt\visitor;
//访问者接口（Visitor Interface）：定义了对各种元素（Element）执行操作的访问者方法。
//具体访问者（Concrete Visitor）：实现访问者接口，并定义对每种类型元素（Element）访问所需进行的具体操作。
//元素接口（Element Interface）：定义了接收访问者的接口（通常是一个 accept 方法）。
//具体元素（Concrete Element）：实现元素接口，通常包含一个 accept 方法，该方法接受一个访问者对象作为参数，并将自身传递给该访问者的对应方法。
//对象结构（Object Structure）：包含元素对象的集合，可以是复合对象或简单的列表，提供一个方法允许访问者访问其内的每个元素。
interface VisitorInterface {
    public function visitConcreteElementA(ConcreteElementA $elementA);
    public function visitConcreteElementB(ConcreteElementB $elementB);
}

interface ElementInterface {
    public function accept(VisitorInterface $visitor);
}

class ConcreteElementA implements ElementInterface {
    public function accept(VisitorInterface $visitor) {
        $visitor->visitConcreteElementA($this);
    }

    public function operationA() {
        // 具体的操作
    }
}

class ConcreteElementB implements ElementInterface {
    public function accept(VisitorInterface $visitor) {
        $visitor->visitConcreteElementB($this);
    }

    public function operationB() {
        // 具体的操作
    }
}

class ConcreteVisitor implements VisitorInterface {
    public function visitConcreteElementA(ConcreteElementA $elementA) {
        $elementA->operationA();
    }

    public function visitConcreteElementB(ConcreteElementB $elementB) {
        $elementB->operationB();
    }
}

// 创建元素对象
$elementA = new ConcreteElementA();
$elementB = new ConcreteElementB();

// 创建访问者
$visitor = new ConcreteVisitor();

// 元素接受访问者
$elementA->accept($visitor);
$elementB->accept($visitor);
